service/src/main/java/org/mskcc/cmo/metadb/service/impl/RequestReplyHandlingServiceImpl.java

```diff
@@ -0,0 +1,162 @@
+    private static volatile boolean shutdownInitiated;
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+    private static final BlockingQueue<ReplyInfo> newRequestReplyQueue =
+        new LinkedBlockingQueue<ReplyInfo>();
+    private static CountDownLatch newRequestHandlerShutdownLatch;
+    private static Gateway messagingGateway;
+    
+    private static final Log LOG = LogFactory.getLog(RequestReplyHandlingServiceImpl.class);
+
+    
+    private class ReplyInfo {
+        String replyMessage;
+        String replyTo;
+        
+        ReplyInfo(String replyMessage, String replyTo) {
+            this.replyMessage = replyMessage;
+            this.replyTo = replyTo;
+        }
+        
+        String getReplyMessage() {
+            return replyMessage;
+        }
+        
+        String getReplyTo() {
+            return replyTo;
+        }
+    }
+    
+    private class NewRequestReplyHandler implements Runnable {
+
+        final Phaser phaser;
+        boolean interrupted = false;
+
+        NewRequestReplyHandler(Phaser phaser) {
+            this.phaser = phaser;
+        }
+
+        @Override
+        public void run() {
+            phaser.arrive();
+            while (true) {
+                try {
+                    ReplyInfo replyInfo = newRequestReplyQueue.poll(100, TimeUnit.MILLISECONDS);
+                    if ((replyInfo != null)) {
+                        System.out.println("Publishing to: " + replyInfo.getReplyTo() + "\n message: "+ replyInfo.getReplyMessage());
+                        messagingGateway.replyPublish(replyInfo.getReplyTo(), replyInfo.getReplyMessage());
+                    } 
+                    if (interrupted && newRequestReplyQueue.isEmpty()) {
+                        break;
+                    }
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                } catch (Exception e) {
+                    LOG.error("Error during request handling", e);
+                }
+            }
+            newRequestHandlerShutdownLatch.countDown();
+        }
+    }
+
+
+    @Override
+    public void initialize(Gateway gateway) throws Exception {
+        if (!initialized) {
+            messagingGateway = gateway;
+            setupPatientSamplesHandler(messagingGateway, this);
+            initializePatientSamplesHandlers();
+            initialized = true;
+        } else {
+            LOG.error("Messaging Handler Service has already been initialized, ignoring patientId.\n");
+        }        
+    }
+
+    @Override
+    public void newPatientSamplesHandler(String patientId, String replyTo) throws Exception {
+        if (!initialized) {
+            throw new IllegalStateException("Message Handling Service has not been initialized");
+        }
+        if (!shutdownInitiated) {
+            List<SampleMetadata> sampleList = sampleService.getSampleMetadataListByCmoPatientId(patientId);
```

Comment:
I would move this into NewRequestReplyHandler.run() so we don't tie up the message handling thread.